design/src/docs/anatomy.md

```diff
@@ -0,0 +1,687 @@
+        layer.layer(inner)
+    }
+```
+
+![Upgradable Diagram](imgs/upgradable-plugin.png)
+
+An example `Plugin` implementation can be found in [aws-smithy-http-server/examples/pokemon-service/src/plugin.rs](https://github.com/awslabs/smithy-rs/blob/main/rust-runtime/aws-smithy-http-server/examples/pokemon-service/src/plugin.rs).
+
+All builders implement the [`Pluggable`](https://github.com/awslabs/smithy-rs/blob/4c5cbc39384f0d949d7693eb87b5853fe72629cd/rust-runtime/aws-smithy-http-server/src/plugin.rs#L8-L29) trait, which allows them to apply plugins to service builders:
+
+```rust
+/// Provides a standard interface for applying [`Plugin`]s to a service builder. This is implemented automatically for
+/// all builders.
+///
+/// As [`Plugin`]s modify the way in which [`Operation`]s are [`upgraded`](crate::operation::Upgradable) we can use
+/// [`Pluggable`] as a foundation to write extension traits which are implemented for all service builders.
+///
+/// # Example
+///
+/// ```
+/// # struct PrintPlugin;
+/// # use aws_smithy_http_server::plugin::Pluggable;
+/// trait PrintExt: Pluggable<PrintPlugin> {
+///     fn print(self) -> Self::Output where Self: Sized {
+///         self.apply(PrintPlugin)
+///     }
+/// }
+///
+/// impl<Builder> PrintExt for Builder where Builder: Pluggable<PrintPlugin> {}
+/// ```
+pub trait Pluggable<NewPlugin> {
+    type Output;
+
+    /// Applies a plugin to the service builder.
+    fn apply(self, plugin: NewPlugin) -> Self::Output;
+}
+```
+
+As seen in the `Pluggable` documentation, third-parties can use extension traits over `Pluggable` to extend the API of builders. In addition to all the `Op{N}` the service builder also holds a `Pl`:
+
+```rust
+/// The service builder for [`PokemonService`].
+///
+/// Constructed via [`PokemonService::builder`].
+pub struct PokemonServiceBuilder<Op1, Op2, Op3, Op4, Op5, Op6, Pl> {
+    capture_pokemon_operation: Op1,
+    empty_operation: Op2,
+    get_pokemon_species: Op3,
+    get_server_statistics: Op4,
+    get_storage: Op5,
+    health_check_operation: Op6,
+
+    plugin: Pl
+}
+```
+
+which allows the following `Pluggable` implementation to be generated:
+
+```rust
+impl<Op1, Op2, /* ... */, Pl, NewPl> aws_smithy_http_server::plugin::Pluggable<NewPl>
+    for PokemonServiceBuilder<Op1, Op2, /* ... */, Pl>
+{
+    type Output =
+        PokemonServiceBuilder<Op1, Exts1, aws_smithy_http_server::plugin::PluginStack<Pl, NewPl>>;
+    fn apply(self, plugin: NewPl) -> Self::Output {
+        PokemonServiceBuilder {
+            capture_pokemon_operation: self.capture_pokemon_operation,
+            empty_operation: self.empty_operation,
+            /* ... */,
+
+            plugin: PluginStack::new(self.plugin, plugin),
+        }
+    }
+}
+```
+
+Here `PluginStack` works in a similar way to [`tower::layer::util::Stack`](https://docs.rs/tower/latest/tower/layer/util/struct.Stack.html) - allowing users to append a new plugin rather than replacing the currently set one.
```

Comment:
It's not immediately clear to me when a user should opt for a `Plugin` instead of attaching a `Layer` to the overall `Router`. 
A concrete example of common functionality that could only be accomplished with a `Plugin` instead of a `Layer` would ⭐

Reply:
That's a very good observation! Plugins allow us to modify the "upgrade procedure" from model services to HTTP services. There we can "squish in" middlewares that work at the model service layer, whereas the ones applied to the overall Router need to operate at the HTTP service layer, _before_ routing has taken place. You'll notice this difference if you look closely at the diagrams.

You can pull the thread for the primary motivation for introducing plugins in item 2 in the [Service Builder Improvements RFC](https://awslabs.github.io/smithy-rs/design/rfcs/rfc0020_service_builder.html#operations-as-middleware-constructors), to perform operation-specific logging behavior, but there are other use cases where we want to modify behavior at the model service layer.

Reply:
How's this?
https://github.com/awslabs/smithy-rs/pull/1782/commits/8ff688e219a6dd41fc73cf65646116d0e657f52f

There's an unfortunate conflict, [which exists for now](https://github.com/awslabs/smithy-rs/issues/1707), between the [`Router`](https://github.com/awslabs/smithy-rs/blob/89a6e5894aebb81efa1305c41188d7ded3b3c43b/rust-runtime/aws-smithy-http-server/src/routing/mod.rs#L64-L67) struct, which is essentially an enum of the [protocol specific routers](https://github.com/awslabs/smithy-rs/blob/89a6e5894aebb81efa1305c41188d7ded3b3c43b/rust-runtime/aws-smithy-http-server/src/routing/mod.rs#L69-L82) and the [`Router`](https://github.com/awslabs/smithy-rs/blob/89a6e5894aebb81efa1305c41188d7ded3b3c43b/rust-runtime/aws-smithy-http-server/src/routers/mod.rs#L40-L47) trait which is an interface for routing requests to services. This makes things more confusing for people today.